### **A Comparative Analysis of Public Sentiment on Twitter towards Apple and Google Products using Natural Language Processing.**  


**Group Name: Group 5**

**Members**
1. *Rose Miriti*
2. *Isaac Wadhare*
3. *Lydia Chumba*
4. *Erick Mauti*
5. *Marilyn Akinyi*
6. *Rodgers Otieno*
7. *Samwel Ongechi*

**Technical Mentor: George Kamundia**

**Phase: Phase 4 Project**


# Sentiment Analysis of Tweets on Apple and Google Products

##  Summary

This project focuses on analyzing public sentiment expressed on Twitter regarding Apple and Google products, using a labeled dataset of over 9,000 tweets categorized as positive, negative, or neutral. The goal is to build a proof-of-concept NLP model capable of classifying tweets according to sentiment, providing actionable insights that can guide business strategy, marketing, and product development for the two companies.

The workflow begins with **business and data understanding**, where the problem is defined, the dataset is explored, and the distribution of sentiment classes is analyzed. The **data preparation** stage includes text cleaning, tokenization, stopword removal, and lemmatization and stemming. Text data will then be transformed into numerical representations using TF-IDF vectors or word embeddings, creating features suitable for machine learning models.

For **modeling**, baseline models such as Logistic Regression and Naive Bayes will be implemented first to establish performance benchmarks. For advanced modeling, we will implement a **neural network**. We will also employ Ensemble methods like Random Forest and Gradient Boosting (XGBoost). 

A **validation strategy** using stratified train-test splits and K-Fold cross-validation will ensure the models generalize well to unseen data. **Evaluation metrics** will include accuracy, precision, recall, F1-score, and confusion matrices to assess multiclass classification performance. The project will produce insights into public sentiment trends, which will directly inform recommendations answering key objectives regarding customer perception, sentiment drivers, and business strategies.


## Business Problem

Apple and Google face continuous public scrutiny on social media regarding product launches and services. Understanding real-time customer sentiment is critical to improve products, marketing strategies, and customer satisfaction.  

**Business problem:**  
*"Can we automatically classify the sentiment of tweets about Apple and Google products to support actionable business insights?"*


## Project Objectives

1. **Determine the overall public sentiment** towards Apple and Google products on Twitter.  
2. **Identify tweet characteristics and themes** that contribute to positive, negative, or neutral sentiment.  
3. **Provide actionable insights** from sentiment trends to inform business decisions, marketing strategies, and product improvements.  

## 1.0 Importing the necessary libraries for the analysis

In [6]:
# Import essential libraries
import nltk
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re
import emoji
import string
from collections import Counter
import warnings
warnings.filterwarnings('ignore')

# Import NLTK specific modules
from nltk.corpus import stopwords, gutenberg
import contractions
from nltk.tokenize import word_tokenize, sent_tokenize, regexp_tokenize
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk import FreqDist, bigrams, trigrams, ngrams
from nltk.collocations import BigramAssocMeasures, BigramCollocationFinder
from nltk import CFG, ChartParser
from nltk import pos_tag
from wordcloud import WordCloud


# Import scikit-learn for machine learning
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV, StratifiedKFold, RandomizedSearchCV
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import ComplementNB
from sklearn.neural_network import MLPClassifier
from sklearn.naive_bayes import MultinomialNB, ComplementNB
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, VotingClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, f1_score, precision_score, recall_score
from sklearn.pipeline import Pipeline
import xgboost as xgb
from xgboost import XGBClassifier
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS, TfidfVectorizer
from sklearn.preprocessing import LabelEncoder
import warnings
warnings.filterwarnings('ignore')
import time


# Set up plotting
plt.style.use('default')
sns.set_palette("husl")

print("All libraries imported successfully!")
print("Environment setup complete!")

from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS

All libraries imported successfully!
Environment setup complete!


## 1.1 Loading the dataset
- Loading the data into a pandas DataFrame and view the first 5 records.

In [7]:
df = pd.read_csv('Data\judge-1377884607_tweet_product_company.csv', encoding = 'latin-1')
df.head()

tweet_text  \
0  .@wesley83 I have a 3G iPhone. After 3 hrs twe...   
1  @jessedee Know about @fludapp ? Awesome iPad/i...   
2  @swonderlin Can not wait for #iPad 2 also. The...   
3  @sxsw I hope this year's festival isn't as cra...   
4  @sxtxstate great stuff on Fri #SXSW: Marissa M...   

  emotion_in_tweet_is_directed_at  \
0                          iPhone   
1              iPad or iPhone App   
2                            iPad   
3              iPad or iPhone App   
4                          Google   

  is_there_an_emotion_directed_at_a_brand_or_product  
0                                   Negative emotion  
1                                   Positive emotion  
2                                   Positive emotion  
3                                   Negative emotion  
4                                   Positive emotion

## 1.2 Data Inspecting 
- Data structure


In [8]:
print("Data Structure")
print(f"Shape: {df.shape} \n")
print(f'{df.info()} \n')  
print("OBSERVATION:")
print(f"There are {df.shape[1]} features. " )
print(f"And {df.shape[0]} records in our dataset." )
print("All features are categorical.")
print("There are missing values in two columns: 'tweet_text' and 'emotion_in_tweet_is_directed_at'.")

Data Structure
Shape: (9093, 3) 



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9093 entries, 0 to 9092
Data columns (total 3 columns):
 #   Column                                              Non-Null Count  Dtype 
---  ------                                              --------------  ----- 
 0   tweet_text                                          9092 non-null   object
 1   emotion_in_tweet_is_directed_at                     3291 non-null   object
 2   is_there_an_emotion_directed_at_a_brand_or_product  9093 non-null   object
dtypes: object(3)
memory usage: 213.2+ KB
None 

OBSERVATION:
There are 3 features. 
And 9093 records in our dataset.
All features are categorical.
There are missing values in two columns: 'tweet_text' and 'emotion_in_tweet_is_directed_at'.


In [9]:
print((df.isnull().sum()/len(df))*100)

tweet_text                                             0.010997
emotion_in_tweet_is_directed_at                       63.807324
is_there_an_emotion_directed_at_a_brand_or_product     0.000000
dtype: float64


### Observation
- **Missing Values**:  
  - `tweet_text`: 0.01%  
  - `emotion_in_tweet_is_directed_at`: 63.8% missing values.  
  - Target column has **no missing values**. 

In [10]:
# Checking for duplicates
print(f"There are {df.duplicated().sum()} duplicates in our dataset. \nWe need to drop them to prevent false outcome .")

There are 22 duplicates in our dataset. 
We need to drop them to prevent false outcome .


- Check uniques values for sentiment features and their frequency

In [11]:
print(f"{df['is_there_an_emotion_directed_at_a_brand_or_product'].value_counts()} \n")

print("Observation:")
print("There is need to merge neutral sentiments and remove the word 'emotion' from positive and negative sentiments")
print("Neutral emotions holds more than 50% of the dataset")


is_there_an_emotion_directed_at_a_brand_or_product
No emotion toward brand or product    5389
Positive emotion                      2978
Negative emotion                       570
I can't tell                           156
Name: count, dtype: int64 

Observation:
There is need to merge neutral sentiments and remove the word 'emotion' from positive and negative sentiments
Neutral emotions holds more than 50% of the dataset


## 2.0 Data Cleaning
**Involves 3 stages:**

### 2.1 Stage 1:
Involves:
- **Column Standardization** - Shortening names for easy reference 
- **Removing duplicates**- Preventing duplicate samples from overweighting certain classes hence every unique tweet target pair appears only once.
- **Mapping sentiment unique values**- Consolidates the neutral emotions and remove emotions from the other parameter

- For this project, we simplify the dataset into three main sentiment classes: 
**positive, negative, and neutral.**  
- **"Positive emotion": Positive**  
- **"Negative emotion" : Negative**  
- **"I can't tell" : Neutral**  
  This reflects uncertainty. It is not clear whether the sentiment is  positive or 
  negative. Mapping it to *neutral* ensures 
  the model doesn’t misinterpret it as an opinion.

- **"No emotion toward brand or product": Neutral**  
  This explicitly signals the absence of any emotional reaction hence it fits best under *neutral*.

In [12]:
# Column Standardization
df = df.rename(columns= {'tweet_text':'text','emotion_in_tweet_is_directed_at':'brand',
                         'is_there_an_emotion_directed_at_a_brand_or_product':'sentiment'})


#Removing duplicates
df = df.drop_duplicates()

#Mapping sentiment unique values
sentiment_mapping = {
    'Negative emotion': 'Negative', 
    'Positive emotion':'Positive',
    'No emotion toward brand or product': 'Neutral',
    "I can't tell":'Neutral'
}
df['sentiment'] = df['sentiment'].map(sentiment_mapping)

### 2.2 Stage 2:
- **Handling missing values**
- Use a dictionary to populate either Apple or Google in all instances where the tweet contains the product names in the list<br>
Index will be used as keys in the dictionary for the sake of matching it with the tweet index later for appropriate categorization<br>
- Replacing null on the brand column with 'Unknown' for the remaining dataset. And dropping null values on the 'text' column the remaining null values

In [13]:
# Dropping null for text column
df = df.dropna(subset=['text'])

# Your brand lists (converted to lowercase for case-insensitive matching)
google_products = [product.lower() for product in [
    "google", "gmail", "gdrive", "google drive", "gdocs", "google docs",
    "gsheets", "google sheets", "gslides", "google slides", "gmeet", "google meet",
    "google calendar", "chrome", "chromebook", "chromecast", "chromium",
    "pixel", "pixelbook", "nest", "google home", "android", "wear os", "play store",
    "google maps", "google earth", "waze", "google photos", "youtube", "yt music",
    "google cloud", "gcp", "google workspace", "google ads", "adsense", "doubleclick"
]]

apple_products = [product.lower() for product in [
    "apple", "icloud", "itunes", "apple music", "apple tv", "apple tv", "apple arcade",
    "apple pay", "apple news", "apple podcasts", "apple watch", "watchos",
    "iphone", "ipad", "ipados", "imac", "macbook", "macbook air", "macbook pro",
    "mac mini", "mac pro", "airpods", "airpods pro", "airpods max",
    "safari", "siri", "facetime", "imessage", "keynote", "numbers", "pages",
    "app store", "ios", "macos", "homepod", "find my", "testflight", "xcode", "iwatch", "mac","macpro"
]]

# Function to identify brand from text
def identify_brand_from_text(text):
    """
    Identify if text contains words related to Google or Apple products.
    Returns 'Apple', 'Google', or 'Unknown'.
    """
    if not isinstance(text, str) or pd.isna(text):
        return 'Unknown'
    
    text_lower = text.lower()
    
    # Check for Apple products
    for product in apple_products:
        if re.search(r'\b' + re.escape(product) + r'\b', text_lower):
            return 'Apple'
    
    # Check for Google products
    for product in google_products:
        if re.search(r'\b' + re.escape(product) + r'\b', text_lower):
            return 'Google'
    
    return 'Unknown'

# Check current brand distribution before imputation
print("Brand distribution BEFORE imputation:")
print(df['brand'].value_counts(dropna=False))
print(f"Missing brands: {df['brand'].isna().sum()}")

# Only impute brands for rows where brand is missing
mask = df['brand'].isna()
df.loc[mask, 'brand'] = df.loc[mask, 'text'].apply(identify_brand_from_text)

# Check brand distribution after imputation
print("\nBrand distribution AFTER imputation:")
print(df['brand'].value_counts(dropna=False))
print(f"Missing brands: {df['brand'].isna().sum()}")

# Check what was imputed
newly_imputed = df[df['brand'].notna() & mask].copy()
print(f"\nNewly imputed brands breakdown:")
print(newly_imputed['brand'].value_counts())

# Show some examples of newly imputed brands
print("\nSample of newly imputed texts:")
for i, row in newly_imputed.head(5).iterrows():
    text_preview = str(row['text'])[:100] + "..." if len(str(row['text'])) > 100 else str(row['text'])
    print(f"Text: {text_preview}")
    print(f"→ Imputed Brand: {row['brand']}")
    print("---")

Brand distribution BEFORE imputation:
brand
NaN                                5788
iPad                                945
Apple                               659
iPad or iPhone App                  469
Google                              428
iPhone                              296
Other Google product or service     293
Android App                          80
Android                              77
Other Apple product or service       35
Name: count, dtype: int64
Missing brands: 5788

Brand distribution AFTER imputation:
brand
Apple                              3834
Google                             2309
iPad                                945
Unknown                             732
iPad or iPhone App                  469
iPhone                              296
Other Google product or service     293
Android App                          80
Android                              77
Other Apple product or service       35
Name: count, dtype: int64
Missing brands: 0

Newly imputed brands